### ISSUE TITLE
Packaging

### ISSUE THREAD

**Issue by GaetanLepage** [SUPERCILEX/clipboard-history#22]
Hi !

I would like to package ringboard on [nixpkgs](https://github.com/NixOS/nixpkgs/).
How would you actually see this software packaged on a distro ?

I ask this question because it is inherently constituted of distinct entities (server, tui client, x11 client, egui...)
Would you have it as a single meta-package shipping all binaries ?
Would each component have its own package that user would install independently ?

Thanks in advance for your help :)

**Comment by SUPERCILEX** [comment]
This is awesome, thanks for looking into packaging!

I'm not sure what the best approach is. I think it could make sense to package everything together in an x11 variant (and in the future wayland). But I can also see people not wanting bloat and preferring individual binaries. If we go that approach, then I think it's important to make sure installing the GUI for example also installs the server and a watcher so people aren't confused. Or maybe we could get the best of both worlds by packaging each binary individually with no dependencies and then offering one package which with nothing in it but dependencies on all the other packages so it acts as an everything package. Thoughts?

**Comment by GaetanLepage** [comment]
Ok, I think I will start by packaging them all individually.
I would still be able to play with "meta-packages" or dependencies later.

There is a blocker for now though, your code seems to rely on a nightly version of Rust. We are only able to compile code with the current stable rustc.
Do you plan to allow this in the future ?

**Comment by SUPERCILEX** [comment]
Ahhhh, I was wondering when the question of nightly would come up. It would be possible to create a patch that makes things work on stable, but it'll be pretty hard. Is it possible to be packaged with some reasonably recent version of nightly? And if things fail to build then you just come yell at me and I'll update to the latest version.

**Comment by msirringhaus** [comment]
I'm also trying to package it for openSUSE at the moment, and am in the same boat: No nightly available in our build-systems. 
There is the **highly** discouraged hack of using `export RUSTC_BOOTSTRAP=1`, which let's one use nightly-features with stable compilers. With this, I can get it to compile, but it will make a lot of people very angry, once they find out.
Having the project rely only on stable features seems to me to be the only non-hacky solution.

**Comment by SUPERCILEX** [comment]
Is there some discussion about why folks would be mad? I think we'd accept that the package will probably break whenever rustc is updated which seems fine as long as the policy is "get rid of the package until they figure out their broken build."

In terms of nightly features, `debug_closure_helpers` would probably be the most painful to remove, though they'll all be a pain to deal with.

**Comment by lolbinarycat** [comment]
> Is there some discussion about why folks would be mad?

`RUST_BOOTSTRAP` is an escape hatch that exists so that `rustc` and the standard library can be built with a stable compiler. it is *not* part of any official api, and it is not intended to be used by anyone except the compiler.  it's existence is already a dirty hack that breaks the rust stability model, using it outside of rustc would be even worse.

**Comment by afifurrohman-id** [comment]
instead of waiting: https://github.com/NixOS/nixpkgs/issues/336643 why not just add `flake.nix` ?

i.e like this: https://github.com/tupakkatapa/mozid/blob/main/flake.nix will make a workaround

**Comment by magnetophon** [comment]
@GaetanLepage Any news on this?
I'd love to be able to install this using nix!

**Comment by magnetophon** [comment]
I made a nix package to build and install the binaries.
It can't be merged upstream yet because it uses nightly.
The services will go in a module, but I haven't started writing that yet.

https://github.com/magnetophon/nixpkgs/blob/ringboard/pkgs/by-name/ri/ringboard/package.nix

**Comment by GaetanLepage** [comment]
Sorry @magnetophon I was quite busy and forgot to reply to your comment.
If you have something working, feel free to open a PR on nixpkgs and tag me :)

**Comment by SUPERCILEX** [comment]
Still using quite a bit: https://github.com/search?q=repo%3ASUPERCILEX%2Fclipboard-history+%22%23%21%5Bfeature%22&type=code

But I think at this point I could come up with polyfills? Do you just need the source code to package ringboard, or would it have to come from crates.io? I could probably create a separate branch that has the features polyfilled so you can build stable versions from source. Does that work?

**Comment by magnetophon** [comment]
That is great news!
Just the source is enough, no crates.io needed.

**Comment by SUPERCILEX** [comment]
Here's what I have for my desktop file:
```
$ cat ~/.local/share/applications/ringboard-egui.desktop
#!/usr/bin/env xdg-open
[Desktop Entry]
Version=1.0
Terminal=false
Type=Application
Name=Ringboard
GenericName=Clipboard Manager
Comment=A fast, efficient, and composable clipboard manager for Linux.
Exec=/bin/sh -c "ps -p `cat /tmp/.ringboard/asaveau.egui-sleep 2> /dev/null` > /dev/null 2>&1 && exec rm -f /tmp/.ringboard/asaveau.egui-sleep || exec /home/asaveau/.cargo/bin/ringboard-egui"
Icon=/home/asaveau/.local/share/icons/hicolor/1024x1024/ringboard.jpeg
StartupWMClass=ringboard-egui
```
So yours looks right to me.

---

> Target mismatch

See https://github.com/systemd/systemd/pull/34606
As for why not always `graphical-session`, the server doesn't depend on using a GUI. It's perfectly valid to add clipboard items from a terminal before you've logged in to a GUI. At least that was my rational, but I have no idea how that would work of if it's useful.

> Service type consistency

AI dumb, obviously I use sd notify.

> Environment logging

Reasonable point, but I compile out all logs lower than info level. The idea is that when you need to debug something, you can build a debug binary that will then have trace level logging and you simply have to restart your systemd services to see what's going on without messing with the service files.

**Comment by magnetophon** [comment]
Thanks for the new release!
Will it be ported to stable as well?

**Comment by SUPERCILEX** [comment]
Yep I forgot. I'm away for the week, but sometime next week I can get to this.

**Comment by magnetophon** [comment]
Great, thanks!  Let me see if I can package that.  :)

Out of curiosity, probably a dumb question: Why do you keep the nightly branch around at all?

**Comment by magnetophon** [comment]
https://github.com/NixOS/nixpkgs/pull/472514

**Comment by SUPERCILEX** [comment]
@magnetophon I just published a new release with the [stable patch](https://github.com/SUPERCILEX/clipboard-history/commit/cb2e94add2388a68a8f015b77f9b082b1658b3b7). Mind sending a nixos PR whenever you have the time?

**Comment by magnetophon** [comment]
Congrats on the new release!
I presume you'll post the stable patch here again?

**Comment by SUPERCILEX** [comment]
Yup, it's already up: https://github.com/SUPERCILEX/clipboard-history/commit/3d95c57bfe939e9f75577101d2aefffc75970cf5

But note that it requires Rust 1.95 which won't be released for another few days.

**Comment by SUPERCILEX** [comment]
Stable patch is up for v0.16.0: https://github.com/SUPERCILEX/clipboard-history/commits/stable/

**Comment by magnetophon** [comment]
Thanks!
When I try to build it, I get:
```
error: build of '/nix/store/adq351d0icy31mpndvw48p3hkl6x3fip-ringboard-0.16.0-unstable-2026-04-27.drv' on 'ssh://nixBuild@builder' failed: Cannot build '/nix/store/adq351d0icy31mpndvw48p3hkl6x3fip-ringboard-0.16.0-unstable-2026-04-27.drv'.
       Reason: builder failed with exit code 101.
       Output paths:
         /nix/store/0dcc3m8sp8fz8wk18wgr5g67b0wh99bx-ringboard-0.16.0-unstable-2026-04-27
       Last 25 log lines:
       >    Compiling crossbeam-channel v0.5.15
       >    Compiling syn v2.0.117
       >    Compiling dirs-sys v0.5.0
       >    Compiling sd-notify v0.5.0
       >    Compiling dirs v6.0.0
       >    Compiling toml_datetime v1.1.1+spec-1.1.0
       >    Compiling serde_spanned v1.1.1
       >    Compiling toml v1.1.2+spec-1.1.0
       >    Compiling thiserror-impl v2.0.18
       >    Compiling serde_derive v1.0.228
       >    Compiling bitcode_derive v0.6.9
       >    Compiling bitcode v0.6.9
       >    Compiling stable-type v0.1.1
       >    Compiling clipboard-history-core v0.16.0 (/build/source/core)
       > error[E0658]: use of unstable library feature `cold_path`
       >   --> core/src/lib.rs:41:13
       >    |
       > 41 |             std::hint::cold_path();
       >    |             ^^^^^^^^^^^^^^^^^^^^
       >    |
       >    = note: see issue #136873 <https://github.com/rust-lang/rust/issues/136873> for more information
       >
       > For more information about this error, try `rustc --explain E0658`.
       > error: could not compile `clipboard-history-core` (lib) due to 1 previous error
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run:
         nix log /nix/store/adq351d0icy31mpndvw48p3hkl6x3fip-ringboard-0.16.0-unstable-2026-04-27.drv
error: Cannot build '/nix/store/adq351d0icy31mpndvw48p3hkl6x3fip-ringboard-0.16.0-unstable-2026-04-27.drv'.
       Reason: builder failed with exit code 1.
       Output paths:
         /nix/store/0dcc3m8sp8fz8wk18wgr5g67b0wh99bx-ringboard-0.16.0-unstable-2026-04-27
       Last 25 log lines:
       >    Compiling crossbeam-channel v0.5.15
       >    Compiling syn v2.0.117
       >    Compiling dirs-sys v0.5.0
       >    Compiling sd-notify v0.5.0
       >    Compiling dirs v6.0.0
       >    Compiling toml_datetime v1.1.1+spec-1.1.0
       >    Compiling serde_spanned v1.1.1
       >    Compiling toml v1.1.2+spec-1.1.0
       >    Compiling thiserror-impl v2.0.18
       >    Compiling serde_derive v1.0.228
       >    Compiling bitcode_derive v0.6.9
       >    Compiling bitcode v0.6.9
       >    Compiling stable-type v0.1.1
       >    Compiling clipboard-history-core v0.16.0 (/build/source/core)
       > error[E0658]: use of unstable library feature `cold_path`
       >   --> core/src/lib.rs:41:13
       >    |
       > 41 |             std::hint::cold_path();
       >    |             ^^^^^^^^^^^^^^^^^^^^
       >    |
       >    = note: see issue #136873 <https://github.com/rust-lang/rust/issues/136873> for more information
       >
       > For more information about this error, try `rustc --explain E0658`.
       > error: could not compile `clipboard-history-core` (lib) due to 1 previous error
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run:
         nix log /nix/store/adq351d0icy31mpndvw48p3hkl6x3fip-ringboard-0.16.0-unstable-2026-04-27.drv
```

**Comment by SUPERCILEX** [comment]
Are you building with rust 1.95?

**Comment by magnetophon** [comment]
Ah, sorry, I should have checked.
Turns out NixOS is still on 1.94.

**Comment by SUPERCILEX** [comment]
Update: the stable patch is now what's being published on crates.io, dunno if that helps with anything. (Nightly is still required to build master.)
